### PR TITLE
If explicit factor is given, don't do conservative checks

### PR DESCRIPTION
**Pull request by abadams**
If we can't prove the explicit factor is ok statically, do dynamic
checks in the inner loop, instead of a conservative check in the outer
loop.

Otherwise we force people to use larger-than-necessary factors depending
on the quality of our bounds visitors, and @dsharletg gets mad.

### CODE REVIEW

**Comment by abadams** [comment]
Fixes #2748

**Comment by steven-johnson** [comment]
Definitely fixes the one case I saw inside Google. Doing some more exhaustive checks with this in place to see if anything else pops up.

**Comment by abadams** [comment]
There may be a performance impact, so you might want to benchmark HDR+ with it

**Comment by steven-johnson** [comment]
@dsharletg reports: "seems like the performance penalty for moving that assert is significant: 70 ms"

**Comment by abadams** [comment]
Blerg. The cost is probably due to the dynamic tracking of the footprint. We have a general problem then. Either the assert is cheap and conservative, or expensive and precise. *How* conservative it is is going to change as the simplifier changes, which means that it rare cases it will increase, as it has done.

**Comment by steven-johnson** [comment]
I had assumed the proper fix was the smarten the simplification logic inside StorageFolding and/or Bounds, so that the compound expression would be simplified down to something equivalent to before.

**Comment by abadams** [comment]
That would be a bandaid. Too much is being asked of that logic.

**Comment by steven-johnson** [comment]
Can't LGTM this at the moment due to perf regression :-(

**Comment by abadams** [comment]
Yeah, not sure how to proceed. We can't gate changes to the simplifier by requiring them to improve performance in every conceivable case. If they improve performance more often than they hurt we should merge them.

We could keep looking at further changes to the simplifier that get this back down, but that's a more fragile approach in general. We could also look at ways of reducing the overhead of the dynamic checks. I don't really have bandwidth for either though.

**Comment by steven-johnson** [comment]
Just touching base on this again so it doesn't drop off our radar. At present, this induces a perf regression that is problematic. We need a long-term fix of some sort; at present, we're (still) hacking out certain simplifier changes in our local copy of Halide to avoid the issue, but that's a path we really don't want to go down in the long term, for obvious reasons.

**Comment by abadams** [comment]
Still on my radar but still no obvious fix. I'm working on making the storage folding prover more powerful in other ways. Maybe it'll help - I'm trying to avoid the selects in the first place.

**Comment by abadams** [comment]
ok, I beefed the analysis up to the point where it can statically determine the value is four for the given example, removing runtime checks entirely. Please benchmark.

**Comment by steven-johnson** [comment]
@dsharletg -- I'll pull and set up a patch if you can point me at the relevant internal benchmarks

**Comment by abadams** [comment]
If you find a different case that fails and can provide a small repro, I'm
happy to keep iterating on this. I need folding to be more robust for my
autoscheduler work.

**Comment by steven-johnson** [comment]
Initial checks look good, doing retesting now

**Comment by steven-johnson** [comment]
the good news: benchmarks look good!
the bad news: some complex tests have nontrivial artifacts. we need to figure out what's up with that.

**Comment by abadams** [comment]
As in incorrect output? Ugh, there must be a subtle problem with one of the
new simplifier rules.

**Comment by steven-johnson** [comment]
@dsharletg suggested running fuzz_simplify for a few hours to see if it catches anything. starting that now.

**Comment by abadams** [comment]
I did one run of that, and am doing it again. Do the failing tests even use
storage folding? That could help narrow it down.

File: src/Bounds.cpp [hunk 1]
```diff
@@ -530,51 +530,149 @@ class Bounds : public IRVisitor {
...
+        op->a.accept(this);
+        Interval a = interval;
+
+        op->b.accept(this);
+        Interval b = interval;
+
+        if (a.is_single_point(op->a) && b.is_single_point(op->b)) {
+            interval = Interval::single_point(op);
+        } else if (a.is_single_point() && b.is_single_point()) {
+            interval = Interval::single_point(a.min && b.min);
```
Comment (steven-johnson) [inline src/Bounds.cpp]: ooops

Comment (abadams) [reply src/Bounds.cpp]: /facepalm

**Comment by steven-johnson** [comment]
LGTM, just waiting on some tests to finish

**Pull request merged**